FV: implement array variable support

rwcarlsen opened the pull request:
This does not yet have support for gradient reconstruction (punting for now).  I still need to write up some markdown docs.

rwcarlsen committed:
fix templating on _ad_u and _ad_grad_u, etc.

These should have always been templated with OutputType before.
Otherwise the functions that work with them have inconsistent return
types among other issues.

rwcarlsen committed:
Add utils/typing details to support AD (fv) array variables

This should also help with an FE array variable implementation
eventually.

rwcarlsen committed:
Create a RealEigenVector instantiation of MooseVariableDataFV.

And create all the necessary member function specializations.  This is
the primary framework update needed to support FV array variables.
Fixes #16074.  This should also pave the way for a smooth implementation
of FE AD array vars (ref #14850).

rwcarlsen committed:
create FVArrayDirichletBC

rwcarlsen committed:
Updating ancilliary interfaces in framework to support (fv) ad array vars.

Including coupleable interface variable warehouse, feproblem, etc.
Needed to add some new related typedefs to moosetypes as well.

rwcarlsen committed:
framework mods to support fv array kernels+bcs

rwcarlsen committed:
add FV array kernels (flux, time, and elemental)

rwcarlsen committed:
remove unnecessary FVElementalKernel::computeOffDiagJacobian

With AD, having these functions split out/separated is not necessary.
Also the FVFluxKernel doesn't have the separate OffDiag func anyway -
this just makes things consistent (and simpler).

rwcarlsen committed:
create shared fv elemental kernel base for array and regular

rwcarlsen committed:
add fv array flux bcs

rwcarlsen committed:
fix bug in AttribVar where non-var objects were matched as equal

rwcarlsen committed:
refix fv dirichlet bcs to implement MooseVariableInterface

Not sure how long this has been a problem - but many queries, etc. on
objects assume we have var-association for fv dirichlet bcs - so we need
to make sure to implement this interface.

rwcarlsen committed:
add AD support to generic constant array prop

rwcarlsen committed:
fix fv kernels and bcs to initialize var interface array-style

rwcarlsen committed:
create shared fv flux kernel base for array and regular

rwcarlsen committed:
add fv array diffusion

rwcarlsen committed:
fix final remaining bugs - FV array simple diffusion works

rwcarlsen committed:
fix IC and time array elemental kernel params bugs

rwcarlsen committed:
add transient fv array kernel test

rwcarlsen committed:
make generic constant 2d array mat support AD

rwcarlsen committed:
add fv array reaction kernel

rwcarlsen committed:
clean up test spec file

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/613005/) on 4562a69 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16073/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16073/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8a08845a97d986075410036714e407eb307ab3e3`

rwcarlsen commented:
Shoot - forgot this will require the libmesh update to get in #16018 before it will compile.

rwcarlsen committed:
moose docs updates for fv array variables

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/614481/) on 1942a40 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16073/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16073/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 30117aad4f743b8efc209820f16e8dc5d441d802`

rwcarlsen committed:
mods to support fv array var coupling right

rwcarlsen committed:
more features to fv array reaction kernel

rwcarlsen committed:
add fv array var coupling test

rwcarlsen committed:
fix diffusion_reaction test consistent with new FV array reaction kernel params

rwcarlsen committed:
fix nasty coupling issue caused by warehouse attrib bug

Previously, the warehouse Var attribute only attempted to cast to
MooseVariableInterface<Real> when reading off the object's variable
number.  This missed indexing/reading the right attribute value (var
number) for RealVectorValue and RealEigenVector templated variables.
Fix this.  This was causing certain queries about whether or not there
were (FV) dirichlet BCs on certain boundaries to come up with no hits
when there were actually dirichlet bc's on that boundary - which caused
jacobians and such to be all messed up.

This also fixes a small quadrature rule bug where we weren't using the
right number of qps in the FV array var when initializing/using it in the FE
loop (e.g. for FE-FV coupling).

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/618017/) on 46b7c7e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/16073/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/16073/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 30117aad4f743b8efc209820f16e8dc5d441d802`

rwcarlsen committed:
fix min gcc compile

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/618178/) on 0431d0a wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/16073/site)

This comment will be updated on new commits.

rwcarlsen committed:
tidy up test spec file

GiudGiud commented:
no no no we need this
